I18n explicit ids

mallex-actee opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
See [[i18n] enable control over translation message ID](https://github.com/angular/angular/issues/12941)


**What is the new behavior?**
Developer can now use specific ID in translations by adding a new "parameter" in the i18n attribute : 
```html
<p i18n="meaning|description@id">foo</p>
<!-- i18n: meaning1|desc1@id1 -->bar<!-- /i18n -->
```

will produce for XMB : 

```xml
<msg id="id" meaning="meaning" description="description">foo</msg>
<msg id="id1" meaning="meaning1" description="description1">foo</msg>
```



**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...


**Other information**:

vsavkin committed:
fix(router): make setUpLocationChangeListener idempotent

vsavkin committed:
fix(router): make setUpLocationChangeListener idempotent

bowenni committed:
refactor(compiler): change ngc error handling

Do not print stack trace for user errors
Print stack trace for compiler internal errors

bowenni committed:
Fix exit code. Give a specific type. Add test cases.

bowenni committed:
Fix format

vicb committed:
refactor(router): cleanup RouterLink

vicb committed:
refactor(router): cleanup RouterLink

vsavkin committed:
fix(router): throw a better error message when angular 1 is not bootstraped

vsavkin committed:
fix(router): throw a better error message when angular 1 is not bootstraped

mallex-actee committed:
feat(compiler): add explicit translation ID option in i18n

Can now add `@id` to i18n meaning and description (`i18n=m|d@i`) to
replace generated ID (sha1 or decimal digest) by this specific ID.

mallex-actee committed:
feat(compiler): add explicit translation ID option in i18n

Can now add `@id` to i18n meaning and description (`i18n=m|d@i`) to
replace generated ID (sha1 or decimal digest) by this specific ID.

mallex-actee committed:
test(compiler) : specific ID in i18n

Testing extracting and merging i18n messages with explicit ID, xliff
and xmb generation.

mallex-actee committed:
test(compiler) : specific ID in i18n

Testing extracting and merging i18n messages with explicit ID, xliff
and xmb generation.

mallex-actee committed:
(feat)compiler : refactor _splitMeaningAndDesc method

Renaming _splitMeaningAndDesc method to _parseI18nAttribute and
refactor function to give more clarity.

mallex-actee committed:
feat(compiler) : refactor _splitMeaningAndDesc method

Renaming _splitMeaningAndDesc method to _parseI18nAttribute and
refactor function to give more clarity.

mallex-actee committed:
Merge remote-tracking branch 'origin/i18n-explicitID' into i18n-explicitID

mallex-actee committed:
style(compiler) format code

Format digest, i18n_parser, digest_spec and extractor_merger_spec

vicb commented:
Thanks for the PR !

I probably won't have time to review before next week.

Meanwhile could you please:
- Make sure that the commit comments follow out guidelines,
- open an issue on angular/angular.io to document this. What should be described is the use case where there is a typo in the source message.

Thanks !

tbosch committed:
fix(compiler): include the summaries of reexported modules / directives / pipes (#13196)

Only if these are not part of the sources.

vicb committed:
fix(router): fix skipLocationChanges on RouterLink directives

fixes #13156

vicb committed:
fix(router): fix skipLocationChanges on RouterLink directives

fixes #13156

vicb committed:
refactor(router): cleanup config

Commit:
fix(router): allow specifying a matcher wihtout specifying a path

fixes #12972

chuckjaz committed:
fix(language-service): update to use `CompilerHost` from compiler-cli (#13189)

IgorMinar committed:
docs(changelog): add instructions for people that depend on the `cache` operator (#13191)

Commit:
docs(core): correct HostListener typo (#13203)

vsavkin committed:
fix(router): runs guards every time when unsuccessfully navigating to the same url over and over again (#13209)

chuckjaz committed:
refactor(build): fix build location of compiler-cli esm module (#13212)

vicb committed:
fix(router): fix replaceUrl on RouterLink directives

vicb committed:
fix(router): fix replaceUrl on RouterLink directives

vicb committed:
refactor(router): code cleanup

vicb committed:
refactor(router): code cleanup

vicb commented on modules/@angular/compiler/src/i18n/digest.ts:
```diff
@@ -9,13 +9,15 @@
 
 export function digest(message: i18n.Message): string {
-  return sha1(serializeNodes(message.nodes).join('') + `[${message.meaning}]`);
+  return (message.id !== '') ?
```
`return message.id || ...`

vicb commented on modules/@angular/compiler/src/i18n/digest.ts:
```diff
@@ -9,13 +9,15 @@
   const visitor = new _SerializerIgnoreIcuExpVisitor();
   const parts = message.nodes.map(a => a.visit(visitor, null));
-  return computeMsgId(parts.join(''), message.meaning);
+  return (message.id !== '') ? message.id : computeMsgId(parts.join(''), message.meaning);
```
same here

vicb commented on modules/@angular/compiler/src/i18n/extractor_merger.ts:
```diff
@@ -337,8 +337,8 @@ class _Visitor implements html.Visitor {
 
-    const [meaning, description] = _splitMeaningAndDesc(meaningAndDesc);
-    const message = this._createI18nMessage(ast, meaning, description);
+    const [meaning, description, id] = _parseI18nAttribute(meaningAndDesc);
```
I would prefer `parsei18nInfo`. Attribute is an impl detail that needs not be exposed here.

vicb commented on modules/@angular/compiler/src/i18n/extractor_merger.ts:
```diff
@@ -496,8 +496,35 @@ function _getI18nAttr(p: html.Element): html.Attribute {
 
-function _splitMeaningAndDesc(i18n: string): [string, string] {
-  if (!i18n) return ['', ''];
+function _parseI18nAttribute(i18n: string): [string, string, string] {
```
open question: change the return type to `{m, d, id}` ?

maxime-allex replied:
Making this change will result on more modifications in the code base (and possibly unwanted side effects) and will not bring anything so I don't think it's a good idea.

vicb commented on modules/@angular/compiler/test/i18n/extractor_merger_spec.ts:
```diff
@@ -29,8 +32,41 @@ export function main() {
+
+      it('should extract from attributes without meaning and with id', () => {
+        expect(extract(
+                   '<div i18n="d1@i1"><span i18n-title="d2@i2" title="single' +
```
remove/improve the line split

vicb commented on modules/@angular/compiler/test/i18n/extractor_merger_spec.ts:
```diff
@@ -29,8 +32,41 @@ export function main() {
+      it('should extract from attributes with ONLY id', () => {
+        expect(extract(
+                   '<div i18n="@i1"><span i18n-title="@i2" title="single' +
+                   ' child">nested</span></div>'))
```
same here

vicb commented on modules/@angular/compiler/test/i18n/integration_spec.ts:
```diff
@@ -92,6 +95,11 @@ export function main() {
       expectHtml(el, '#i18n-15').toMatch(/ca <b>devrait<\/b> marcher/);
+
+      expectHtml(el, '#i18n-16').toMatch(/avec un ID explicite/);
+
```
revert empty lines

vicb commented:
looks mostly good to me after a quick review.
I'll looker more deeply next week.
Thanks for the PR !

maxime-allex commented on modules/@angular/compiler/src/i18n/extractor_merger.ts:
```diff
@@ -496,8 +496,35 @@ function _getI18nAttr(p: html.Element): html.Attribute {
+   *    i18n="@m|d"   => ['@m', 'd', '']
+   *    i18n="@i"     => ['', '', 'i']
+   *    i18n="d@i"    => ['', 'd', 'i']
+   */
```
> Probably it's too smart to explain it in the docs

I don't think users need to know (in the docs) implementation details. However the rule is quiet simple : 
* everything before the first `|` is the meaning
* everything after the last `@` is the ID
* the `|` take precedence other the `@` so user can use `@` in meaning - or in description if an ID is set

> I would say everything after @ is an id and maybe provide a way to escape @ (@@ ?)

I'm not for creating special escaping things, it will complicate unnecessarily integration and maintainability.

mallex-actee committed:
Merge remote-tracking branch 'angular/master' into i18n-explicitID

mallex-actee committed:
refactor(compiler): rename _parseI18nAttribute method

renaming _parseI18nAttribute method to _parsei18nInfo and renaming
meaningAndDesc value to i18nInfo.

mallex-actee committed:
refactor(compiler): use logical operator in digest and decimalDigest methods

Use logical operator rather than conditional one in digest and
decimalDigest methods

mallex-actee committed:
refactor(compiler): use logical operator in digest and decimalDigest methods

Use logical operator rather than conditional one in digest and
decimalDigest methods

mallex-actee committed:
style(compiler): review line splitting

refactor line splitting in extractor_merger_spec and integration_spec
tests suites.

mallex-actee committed:
style(compiler): review line splitting

refactor line splitting in extractor_merger_spec and integration_spec
tests suites.

belev committed:
docs(contributing): fix styleguide link and typos (#13198)

belev committed:
docs(contributing): fix styleguide link and typos (#13198)

Commit:
docs(cheatsheet): correct HostBinding decorator example (#13206)

Commit:
docs(cheatsheet): correct HostBinding decorator example (#13206)

awerlang committed:
fix(facade): cache original format string (#12764)

awerlang committed:
fix(facade): cache original format string (#12764)

belev committed:
docs(developer): add anchor to source code formatting (#13199)

In CONTRIBUTING.md#rules has reference to DEVELOPER.md#clang-format
but anchor to #clang-format is not present

belev committed:
docs(developer): add anchor to source code formatting (#13199)

In CONTRIBUTING.md#rules has reference to DEVELOPER.md#clang-format
but anchor to #clang-format is not present

tbosch committed:
refactor(compiler): replace `isStaticSymbol` with `instanceof StaticSymbol`

tbosch committed:
refactor(compiler): replace `isStaticSymbol` with `instanceof StaticSymbol`

tbosch committed:
fix(compiler): serialize any `StaticSymbol` correctly, not matter in which context.

tbosch committed:
fix(compiler): serialize any `StaticSymbol` correctly, not matter in which context.

vicb commented:
Could you please address the last comments, rebase and squash your commits (by logical change) ?

Thanks

mallex-actee committed:
# This is a combination of 3 commits.
# The first commit's message is:
(feat)compiler : explicit ID now prefixed with '@@'

# This is the 2nd commit message:

refactor(compiler): use logical operator in digest and decimalDigest methods

Use logical operator rather than conditional one in digest and
decimalDigest methods

# This is the 3rd commit message:

refactor(compiler): rename _parseI18nAttribute method

renaming _parseI18nAttribute method to _parsei18nInfo and renaming
meaningAndDesc value to i18nInfo.

mallex-actee committed:
test(compiler) : update tests for new '@@' prefix

test(compiler) : specific ID in i18n

Testing extracting and merging i18n messages with explicit ID, xliff
and xmb generation.

mallex-actee committed:
test(compiler) : update tests for new '@@' prefix

test(compiler) : specific ID in i18n

Testing extracting and merging i18n messages with explicit ID, xliff
and xmb generation.

mallex-actee committed:
style(compiler): review extractor_merger

mallex-actee committed:
test(compiler): review tests line splitting

mallex-actee committed:
test(compiler): review tests line splitting

mallex-actee committed:
Merge remote-tracking branch 'origin/i18n-explicitID' into i18n-explicitID

DzmitryShylovich commented:
You need to `rebase` your branch:
```
git fetch upstream
git rebase upstream/master
```

maxime-allex commented:
@DzmitryShylovich Thanks !
